pilot/pkg/networking/core/configgen.go

```diff
@@ -42,6 +43,9 @@ type ConfigGenerator interface {
 	// BuildNameTable returns list of hostnames and the associated IPs
 	BuildNameTable(node *model.Proxy, push *model.PushContext) *nds.NameTable
 
+	// BuildExtensionConfiguration returns the list of extension configuration for the given proxy and list of names. This is the ECDS output.
+	BuildExtensionConfiguration(node *model.Proxy, push *model.PushContext, extensionConfigNames []string) []*core.TypedExtensionConfig
```

Comment:
I'm not familiar with the xds code. What version is assigned to resources? It would be better to not shuffle versions  for no reason since it's costly in Envoy.

Reply:
We assign a version based on a global version incremented each time PushContext is re-initialized (ie every config change). The same version is used for all resources/types

Reply:
Is the version same for all resources? That means update to any will cause update to all?

Reply:
Yes. Isn't version in DiscoveryResponse so that is necessary? 

I suppose we could have per-type version, but we do not today. However, for ECDS, we get effectively the same behavior. We only push to envoy if EnvoyFilter has changed, so we only push when we would have to change the version anyways

Reply:
Ideally, the version is per resource (name and type). If any EnvoyFilter causes a bump in version for all ECDS resources, then that would be not scalable. But it's not a problem if we don't expect that many EnvoyFilters.

Reply:
Ah, I see with https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/discovery.proto#envoy-api-msg-resource you can now do it per resource. I think previously you couldn't?

Isn't envoy doing a hash of the config for comparison anyways? Does version matching short-circuit the hashing?

Reply:
We have the same issue for all resource types, which are orders of magnitude more than ECDS probably, so this is likely something we need to look into in general but may not matter much for ECDS

Reply:
Per-resource versions were there since the beginning. This is a bigger problem for ECDS since 1) inline bytes will cost a lot to hash, 2) inline file names do not participate in hashing (likely causes of bugs like in SDS).

Reply:
> inline file names do not participate in hashing (likely causes of bugs like in SDS).

Why it is not participating in hashing?

Reply:
File contents are not part of the resource. I think this is fine with the current design since the file name is the hash, so it should trigger an update when needed.